bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/ExtendedDiscoveryServiceTracker.java

```diff
@@ -0,0 +1,155 @@
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingRegistry;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link ExtendedDiscoveryServiceTracker} adds the {@link DiscoveryServiceCallback} to all
+ * {@link ExtendedDiscoveryService} instances.
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true)
+public class ExtendedDiscoveryServiceTracker {
+
+    private @Nullable Inbox inbox;
+    private @Nullable ThingRegistry thingRegistry;
+    private @Nullable DiscoveryServiceCallback discoveryServiceCallback;
+
+    private final List<ExtendedDiscoveryService> extendedDiscoveryServices = new ArrayList<>();
+    private final AtomicBoolean active = new AtomicBoolean(false);
+
+    @Reference
+    protected void addDiscoveryService(DiscoveryService discoveryService) {
+        if (discoveryService instanceof ExtendedDiscoveryService) {
+            ExtendedDiscoveryService extendedDiscoveryService = (ExtendedDiscoveryService) discoveryService;
+            extendedDiscoveryServices.add(extendedDiscoveryService);
+
+            if (active.get()) {
+                DiscoveryServiceCallback dsc = this.discoveryServiceCallback;
+                if (dsc != null) {
+                    extendedDiscoveryService.setDiscoveryServiceCallback(dsc);
+                }
+            }
+        }
+    }
+
+    protected void removeDiscoveryService(DiscoveryService discoveryService) {
+        if (discoveryService instanceof ExtendedDiscoveryService) {
+            ExtendedDiscoveryService extendedDiscoveryService = (ExtendedDiscoveryService) discoveryService;
+            extendedDiscoveryServices.remove(extendedDiscoveryService);
+        }
+    }
+
+    @Reference
+    protected void setInbox(Inbox inbox) {
+        this.inbox = inbox;
+    }
+
+    protected void unsetInbox(Inbox inbox) {
+        this.inbox = null;
+    }
+
+    @Reference
+    protected void setThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = thingRegistry;
+    }
+
+    protected void unsetThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = null;
+    }
+
+    @Activate
+    protected void activate() {
+        DiscoveryServiceCallback dsc = createDiscoveryServiceCallback();
+        this.discoveryServiceCallback = dsc;
+        active.set(true);
+
+        for (ExtendedDiscoveryService extendedDiscoveryService : extendedDiscoveryServices) {
+            extendedDiscoveryService.setDiscoveryServiceCallback(dsc);
+        }
+
+    }
+
+    @Deactivate
+    protected void deactivate() {
+        active.set(false);
+        extendedDiscoveryServices.clear();
```

Comment:
Hm, is it allowed that the extended discovery services still use the callback if this component has been deactivated (e.g. because a mandatory reference gets lost)?
Shouldn't you unset the callback?

Reply:
I though about this, too. Its currently marked `@NonNull` so this would change the API. As this is an interface out consumers implement its not nice to pass `null` all of a sudden
An alternative would be to unset the inbox & thingRegistry fields from the callback implementation and safe guard the calls to them. wdyt?

Reply:
You could create one static reference (e.g. disabledDsc) of a anonymous `DiscoveryServiceCallback` implementation, thats `getExistingThing` and `getExistingDiscoveryResult` return null.

Then you can keep mutable objects and just need to update the callback (non null) to the services.